### PR TITLE
FEATURE: impersonation now requires second factor authorization

### PR DESCRIPTION
**Pull request by techAPJ**
Screenshot 2021-04-09 at 14 37 54
Screenshot 2021-04-09 at 14 38 04

### CODE REVIEW

**Comment by ZogStriP** [comment]
Quick question: why is the `@irish_kozey` "mention" styled differently in the 1st modal?

File: app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  @action
+  impersonate() {
+    ajax("/admin/impersonate", {
+      type: "POST",
+      data: {
+        username_or_email: this.username,
+        second_factor_token:
+          this.securityKeyCredential || this.secondFactorToken,
+        second_factor_method: this.secondFactorMethod,
+        timezone: moment.tz.guess(),
```
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js:49]: Why is the timezone required?

File: app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      },
+    }).then(
+      (result) => {
+        // Successful login
+        if (result && result.error) {
```
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js:54]: Is it a successful login if there's an error?

File: app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+              securityKeyAllowedCredentialIds: result.allowed_credential_ids,
+            });
+
+            // only need to focus the 2FA input for TOTP
+            if (!this.showSecurityKey) {
+              schedule("afterRender", () =>
+                document
+                  .getElementById("second-factor")
+                  .querySelector("input")
+                  .focus()
```
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/controllers/modals/admin-impersonate-user.js:79]: You could shorten that a bit.

```suggestion
                document.querySelector("#second-factor input").focus()
```

**Comment by SamSaffron** [comment]
sorry @techAPJ I am closing this, when you free up you can spend some time with @martin-brennan doing the "generic" reusable 2fa screen. Then it should be a breeze plugging this in anywhere we need it.

**Pull request closed without merging**